### DISCUSSION TITLE
How to dynamically update services and routing configuration in kong dbless mode?

### DISCUSSION THREAD

**Discussion by zyiqian** [Kong/kong#12821] [Help] [answered]
Hi everyone..

### [my question]
I will disable the Ingress Controller and enabled dblessConfig. When deployed to k8s, only one proxy container will be run.

When I need to configure the services and routing in dblessConfig.config. Is it possible to update dynamically without restarting the proxy container?

Because if the proxy container is restarted, it will affect our online long-connection service. We don't want to be disconnected and reconnected.

We now update dynamically through Ingress.But I want to switch to dbless mode. I don’t know if it can have the same effect.

I don't have any idea for resolve task. Can you help me? thank you

### My KONG setup
- kong:3.1

**Comment by chronolaw** [comment]
I think that the answer is yes, kong can always  update the configuration without reloading.

**Reply by zyiqian** [reply]
hi @chronolaw ,

Sorry for the late reply, I want to know how to configure kong, or is there any documentation for me to learn from? Because I searched for a long time and didn't find it. thank you.

**Reply by chronolaw** [reply]
This may be helpful .https://docs.konghq.com/gateway/latest/admin-api/

**Reply by zyiqian** [reply]
I tried to create a service and route according to this guide, and the following error was returned

`$ curl -i -s -X POST http://localhost:8001/services \
 --data name=test-service \
 --data url='http://httpbin.org'
HTTP/1.1 405 Not Allowed
Date: Sun, 07 Apr 2024 03:33:00 GMT
Content-Type: application/json; charset=utf-8
Connection: keep-alive
Access-Control-Allow-Origin: *
Content-Length: 114
X-Kong-Admin-Latency: 5

{"code":12,"message":"cannot create 'services' entities when not using a database","name":"operation unsupported"}%

$ curl -i -X POST http://localhost:8001/services/test-service/routes \
    -H "Content-Type: application/json" \
    -d '{"name": "test-service1-route", "paths": [ "/path/one", "/path/two" ]}'
HTTP/1.1 405 Not Allowed
Date: Sun, 07 Apr 2024 03:35:16 GMT
Content-Type: application/json; charset=utf-8
Connection: keep-alive
Access-Control-Allow-Origin: *
Content-Length: 112
X-Kong-Admin-Latency: 6

{"code":12,"message":"cannot create 'routes' entities when not using a database","name":"operation unsupported"}%`

**Reply by chronolaw** [reply, marked as answer]
This is declarative configuration's help page: https://docs.konghq.com/gateway/latest/admin-api/declarative-configuration/.

You must generate a YAML file first, then use the `/config` endpoint, like:
```
curl 127.1:8001/config -F config=@kong.yml | jq
```

**Reply by zyiqian** [reply]
Great. I can use this method to dynamically update service and route without restarting kong. Thank you very much.

```
$ ls
kong.yml
$ curl -i -X POST http://localhost:8001/config \
 --form config=@kong.yml
 
 
 HTTP/1.1 201 Created
 Date: Tue, 09 Apr 2024 02:32:03 GMT
 ......
```
